fuzzypp/src/preprocessor.cpp

```diff
@@ -0,0 +1,60 @@
+#include <algorithm>
+#include <filesystem>
+#include <fstream>
+
+#include <iostream>
+
+#include "preprocessor.hpp"
+
+namespace fuzzypp::preprocessor {
+    void 
+    FuzzyPreprocessor::preprocess(const fuzzypp::cliopts::CliOptions& options) {
+        auto opts = generate_simplecpp_opts(options);
+        const auto output_path = std::filesystem::path { options.output_directory };
+
+        auto write_to_file = [&](const auto& filename) {
+            auto file_path = std::filesystem::path { filename }.relative_path();
+            auto output_file = (std::filesystem::weakly_canonical(output_path) /= file_path).lexically_normal();
+            if (output_file.has_parent_path()) std::filesystem::create_directories(output_file.parent_path());
+            
+            std::ofstream output { output_file, std::ofstream::trunc };
+            output << stringify(filename, opts);
+        };
+
+        std::for_each(options.files.cbegin(),
+                      options.files.cend(),
+                      write_to_file);
+    }
+
+    const simplecpp::DUI
+    FuzzyPreprocessor::generate_simplecpp_opts(const fuzzypp::cliopts::CliOptions& options) {
+        simplecpp::DUI simple_opts;
+
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
+        simple_opts.includePaths = std::list<std::string> { options.include_paths.cbegin(), options.include_paths.cend() };
+        simple_opts.defines = std::list<std::string> { options.defines.cbegin(), options.defines.cend() };
+        simple_opts.undefined = std::set<std::string> { options.undefines.cbegin(), options.undefines.cend() };
```

Comment:
`undefined` x `undefines`

Reply:
As with the `includePaths` - stuck with this name :(